src/main/java/io/cryostat/recordings/RecordingArchiveHelper.java

```diff
@@ -39,90 +39,244 @@
+                                        })
+                                .filter(Objects::nonNull)
+                                .collect(Collectors.toList());
+                archivedRecordings.addAll(temp);
+            }
+            future.complete(archivedRecordings);
+        } catch (ArchivePathException e) {
+            future.completeExceptionally(e);
+        }
+
+        return future;
+    }
+
+    public Future<Path> getRecordingPath(String recordingName) throws Exception {
+        CompletableFuture<Path> future = new CompletableFuture<>();
 
-                            if (descriptor.isPresent()) {
-                                return writeRecordingToDestination(connection, descriptor.get());
-                            } else {
-                                throw new RecordingNotFoundException(recordingName);
-                            }
-                        });
+        try {
+            List<String> subdirectories = this.fs.listDirectoryChildren(recordingsPath);
+            Path archivedRecording = searchSubdirectories(subdirectories, recordingsPath, recordingName);
+            if (!(fs.exists(archivedRecording)
+                    && fs.isRegularFile(archivedRecording)
+                    && fs.isReadable(archivedRecording))) {
+                throw new RecordingNotFoundException(recordingName);
+            }
+            future.complete(archivedRecording);
+        } catch (RecordingNotFoundException e) {
+            future.completeExceptionally(e);
+        }
 
-        return saveName;
+        return future;
     }
 
-    public void deleteRecording(ConnectionDescriptor connectionDescriptor, String recordingName)
-            throws Exception {
-
-        targetConnectionManager.executeConnectedTask(
-                connectionDescriptor,
-                connection -> {
-                    Optional<IRecordingDescriptor> descriptor =
-                            this.getDescriptorByName(connection, recordingName);
-
-                    if (descriptor.isPresent()) {
-                        connection.getService().close(descriptor.get());
-                        reportService.delete(connectionDescriptor, recordingName);
-                    } else {
-                        throw new RecordingNotFoundException(recordingName);
-                    }
-                    return null;
-                });
+    private Path searchSubdirectories(List<String> subdirectories, Path parent, String recordingName) throws Exception {
+        Path archivedRecording = null;
+        for (String subdirectory : subdirectories) {
+            List<String> files =
+                    this.fs.listDirectoryChildren(parent.resolve(subdirectory));
+
+            for (String file : files) {
+                if (recordingName.equals(file)) {
+                    archivedRecording = parent.resolve(subdirectory).resolve(file).normalize().toAbsolutePath();
+                    return archivedRecording;
```

Comment:
No need to assign the `archivedRecording` value and then also immediately return it here - you can just directly return the found value.

In fact, you could remove the `archivedRecording` local variable altogether, and simply return the result here. If you get to the end of the method without having returned anything then you can either `return null` or `throw RecordingNotFoundException`.